File: frontend/lib/api/images.ts
/**
 * Image upload API client
 */

import axios from 'axios';
import { getApiBaseUrl } from '../config';

const API_BASE_URL = getApiBaseUrl();

export interface UploadImageResponse {
  url: string;
  message: string;
}

export interface UploadMultipleImagesResponse {
  urls: string[];
  errors?: string[] | null;
  message: string;
}

/**
 * Upload a single image file
 */
export async function uploadImage(
  file: File,
  authToken: string
): Promise<UploadImageResponse> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await axios.post<UploadImageResponse>(
    `${API_BASE_URL}/api/v1/images/upload`,
    formData,
    {
      headers: {
        'Authorization': `Bearer ${authToken}`,
        'Content-Type': 'multipart/form-data',
      },
    }
  );

  return response.data;
}

/**
 * Upload multiple image files (max 5)
 */
export async function uploadMultipleImages(
  files: File[],
  authToken: string
): Promise<UploadMultipleImagesResponse> {
  if (files.length > 5) {
    throw new Error('Maximum 5 images allowed per request');
  }

  const formData = new FormData();
  files.forEach((file) => {
    formData.append('files', file);
  });

  const response = await axios.post<UploadMultipleImagesResponse>(
    `${API_BASE_URL}/api/v1/images/upload-multiple`,
    formData,
    {
      headers: {
        'Authorization': `Bearer ${authToken}`,
        'Content-Type': 'multipart/form-data',
      },
    }
  );

  return response.data;
}

/**
 * Delete an image from storage
 */
export async function deleteImage(
  imageUrl: string,
  authToken: string
): Promise<void> {
  await axios.delete(`${API_BASE_URL}/api/v1/images/delete`, {
    headers: {
      'Authorization': `Bearer ${authToken}`,
    },
    params: {
      image_url: imageUrl,
    },
  });
}

/**
 * Validate image file
 */
export function validateImageFile(file: File): { valid: boolean; error?: string } {
  // Check file type
  const allowedTypes = ['image/jpeg', 'image/jpg', 'image/png', 'image/gif', 'image/webp'];
  if (!allowedTypes.includes(file.type)) {
    return {
      valid: false,
      error: 'Invalid file type. Allowed types: JPEG, PNG, GIF, WebP',
    };
  }

  // Check file size (10MB max)
  const maxSize = 10 * 1024 * 1024; // 10MB
  if (file.size > maxSize) {
    return {
      valid: false,
      error: 'File size exceeds maximum allowed size of 10MB',
    };
  }

  return { valid: true };
}

/**
 * Create a preview URL for an image file
 */
export function createImagePreview(file: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = (e) => {
      if (e.target?.result) {
        resolve(e.target.result as string);
      } else {
        reject(new Error('Failed to create preview'));
      }
    };
    reader.onerror = () => reject(new Error('Failed to read file'));
    reader.readAsDataURL(file);
  });
}

File: backend/app/api/images.py
"""
Image upload API endpoints for Google Cloud Storage.
"""

import logging
from typing import List

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile, status
from fastapi.responses import JSONResponse

from app.auth.dependencies import get_current_user
from app.services.storage_service import storage_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/images", tags=["images"])

# Allowed image MIME types
ALLOWED_CONTENT_TYPES = {
    "image/jpeg",
    "image/jpg",
    "image/png",
    "image/gif",
    "image/webp",
}

# Maximum file size: 10MB
MAX_FILE_SIZE = 10 * 1024 * 1024


@router.post(
    "/upload",
    summary="Upload an image to Google Cloud Storage",
    response_class=JSONResponse,
)
async def upload_image(
    file: UploadFile = File(...),
    current_user: dict = Depends(get_current_user),
):
    """
    Upload an image file to Google Cloud Storage.
    
    Returns the public URL of the uploaded image.
    
    Requires authentication.
    """
    # Check if storage service is available
    if not storage_service.is_available():
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Image upload service is not configured. Please set GCS_BUCKET_NAME and GCS_PROJECT_ID environment variables.",
        )

    # Validate content type
    if file.content_type not in ALLOWED_CONTENT_TYPES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid file type. Allowed types: {', '.join(ALLOWED_CONTENT_TYPES)}",
        )

    # Read file content
    try:
        image_data = await file.read()
    except Exception as e:
        logger.error(f"Failed to read uploaded file: {e}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Failed to read uploaded file",
        )

    # Check file size
    if len(image_data) > MAX_FILE_SIZE:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"File size exceeds maximum allowed size of {MAX_FILE_SIZE / 1024 / 1024}MB",
        )

    # Upload to GCS
    image_url = storage_service.upload_image(
        image_data=image_data,
        content_type=file.content_type or "image/jpeg",
        folder="posts",
        resize=True,
    )

    if not image_url:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to upload image to storage",
        )

    return JSONResponse(
        status_code=status.HTTP_201_CREATED,
        content={
            "url": image_url,
            "message": "Image uploaded successfully",
        },
    )


@router.post(
    "/upload-multiple",
    summary="Upload multiple images to Google Cloud Storage",
    response_class=JSONResponse,
)
async def upload_multiple_images(
    files: List[UploadFile] = File(...),
    current_user: dict = Depends(get_current_user),
):
    """
    Upload multiple image files to Google Cloud Storage.
    
    Returns a list of public URLs of the uploaded images.
    
    Maximum 5 images per request.
    
    Requires authentication.
    """
    # Check if storage service is available
    if not storage_service.is_available():
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Image upload service is not configured. Please set GCS_BUCKET_NAME and GCS_PROJECT_ID environment variables.",
        )

    # Validate number of files
    if len(files) > 5:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Maximum 5 images allowed per request",
        )

    uploaded_urls = []
    errors = []

    for file in files:
        # Validate content type
        if file.content_type not in ALLOWED_CONTENT_TYPES:
            errors.append(f"{file.filename}: Invalid file type")
            continue

        # Read file content
        try:
            image_data = await file.read()
        except Exception as e:
            logger.error(f"Failed to read uploaded file {file.filename}: {e}")
            errors.append(f"{file.filename}: Failed to read file")
            continue

        # Check file size
        if len(image_data) > MAX_FILE_SIZE:
            errors.append(f"{file.filename}: File size exceeds maximum allowed size")
            continue

        # Upload to GCS
        image_url = storage_service.upload_image(
            image_data=image_data,
            content_type=file.content_type or "image/jpeg",
            folder="posts",
            resize=True,
        )

        if image_url:
            uploaded_urls.append(image_url)
        else:
            errors.append(f"{file.filename}: Failed to upload")

    if not uploaded_urls and errors:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to upload images: {', '.join(errors)}",
        )

    return JSONResponse(
        status_code=status.HTTP_201_CREATED,
        content={
            "urls": uploaded_urls,
            "errors": errors if errors else None,
            "message": f"Successfully uploaded {len(uploaded_urls)} image(s)",
        },
    )


@router.delete(
    "/delete",
    summary="Delete an image from Google Cloud Storage",
    response_class=JSONResponse,
)
async def delete_image(
    image_url: str,
    current_user: dict = Depends(get_current_user),
):
    """
    Delete an image from Google Cloud Storage.
    
    Requires authentication.
    """
    # Check if storage service is available
    if not storage_service.is_available():
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Image upload service is not configured",
        )

    success = storage_service.delete_image(image_url)

    if not success:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to delete image",
        )

    return JSONResponse(
        status_code=status.HTTP_200_OK,
        content={
            "message": "Image deleted successfully",
        },
    )

